### PR TITLE
feat(epmodel): Event Processor datamodel

### PR DESCRIPTION
**Pull request by msohailhussain**
## Summary
- Data models to represent Impression/Conversion event on client/server side.

## Test plan
- Unit tests to verify after serialization datamodel is converted into expected server payload
Impression Event
Conversion Event

### CODE REVIEW

File: lib/optimizely/event/entity/conversion_event.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+require_relative 'user_event'
+module Optimizely
+  class ConversionEvent < UserEvent
```
Comment (msohailhussain) [inline lib/optimizely/event/entity/conversion_event.rb]: where are other attributes? 
 bot filter?

File: lib/optimizely/event/entity/conversion_event.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    attr_accessor :event, :event_tags
```
Comment (msohailhussain) [inline lib/optimizely/event/entity/conversion_event.rb]: can't we just define attr_reader and pass it in a constructor?

File: lib/optimizely/event/entity/event_batch.rb [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        revision: @revision,
+        client_name: @client_name,
+        client_version: @client_version,
+        anonymize_ip: @anonymize_ip,
+        enrich_decisions: @enrich_decisions,
+        visitors: @visitors
+      }
+    end
+
+    class Builder
```
Comment (msohailhussain) [inline lib/optimizely/event/entity/event_batch.rb]: do we really need builder classes in ruby ?

File: lib/optimizely/event/entity/event_context.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+module Optimizely
+  class EventContext
+    attr_reader :account_id, :project_id, :revision, :client_name,
```
Comment (msohailhussain) [inline lib/optimizely/event/entity/event_context.rb]: how you will set these values?

File: lib/optimizely/event/entity/impression_event.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+require_relative 'user_event'
+module Optimizely
+  class ImpressionEvent < UserEvent
+    attr_reader :user_id, :user_attributes
```
Comment (msohailhussain) [inline lib/optimizely/event/entity/impression_event.rb]: it's not user_attributes, it's visitor_attributes

File: lib/optimizely/event/entity/impression_event.rb [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    attr_accessor :experiment, :variation
```
Comment (msohailhussain) [inline lib/optimizely/event/entity/impression_event.rb]: why accessor?

File: lib/optimizely/event/entity/snapshot_event.rb [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+      value: nil,
+      event_tags: nil
+    )
+      @entity_id = entity_id
+      @uuid = uuid
+      @key = key
+      @timestamp = timestamp
+      @revenue = revenue
+      @value = value
+      @event_tags = event_tags
```
Comment (msohailhussain) [inline lib/optimizely/event/entity/snapshot_event.rb]: these are just tags?

File: lib/optimizely/event/entity/user_event.rb [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+module Optimizely
+  class UserEvent
+    attr_reader :context, :uuid, :timestamp
```
Comment (msohailhussain) [inline lib/optimizely/event/entity/user_event.rb]: write it as event_context

File: lib/optimizely/event/entity/event_batch.rb (hunk 3 above)
Comment (rashidsp) [reply lib/optimizely/event/entity/event_batch.rb]: Builder is being used to handle the multi-part objects.
However can simply initialize the nested attributes as well.

File: lib/optimizely/event/entity/impression_event.rb (hunk 6 above)
Comment (rashidsp) [reply lib/optimizely/event/entity/impression_event.rb]: Following csharp-sdk.
https://github.com/optimizely/csharp-sdk/blob/master/OptimizelySDK/Event/Entity/ImpressionEvent.cs#L30

File: lib/optimizely/event/entity/event_context.rb (hunk 4 above)
Comment (rashidsp) [reply lib/optimizely/event/entity/event_context.rb]: initialized

**Pull request merged**